### ISSUE TITLE
Calling generic constructor in method returning `Self` generates partially unknown error

### ISSUE THREAD

**Issue by bryanforbes** [microsoft/pyright#3765]
**Describe the bug**
Given the following code:

```python
from __future__ import annotations

from typing import Any, Generic, Protocol, TypeVar
from typing_extensions import Self

T_contra = TypeVar('T_contra', contravariant=True)
ThingT = TypeVar('ThingT', bound='Thing[Any]')


class Callback(Protocol[T_contra]):
    def __call__(self, message: T_contra, /) -> Any:
        ...


class Thing(Generic[T_contra]):
    def __init__(self, callback: Callback[T_contra]) -> None:
        self._callback: Callback[T_contra] = callback

    def copy(self) -> Self:
        return Thing(self._callback)
```

Calling the constructor in `Thing.copy()` reports that the return type is partially unknown.

**Expected behavior**
No error is generated.

**VS Code extension or command-line**
Are you running pyright as a VS Code extension or a command-line tool?
Command line

Which version? 1.1.264

**Additional context**
Using the bound `TypeVar` from line 7 instead of `Self` (`def copy(self: ThingT) -> ThingT:`) results in an error that says `Expression of type "Thing[Any]" cannot be assigned to return type "ThingT@copy"`.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.265, which I just published. It will be included in a future release of pylance.
